zwave_js_server/util/node.py

```diff
@@ -115,6 +74,158 @@ async def async_set_config_parameter(
+
+
+async def async_bulk_set_partial_config_parameters(
+    node: Node,
+    new_value: Union[int, Dict[int, Union[int, str]]],
+    property_: int,
+) -> None:
+    """Bulk set partial configuration values on this node."""
+    config_values = node.get_configuration_values()
+    property_values = [
+        value for value in config_values.values() if value.property_ == property_
+    ]
+
+    # If we can't find any values with this property, the property is wrong
+    if len(property_values) == 0:
+        raise NotFoundError(
+            f"Configuration parameter {property_} for node {node.node_id} not found"
+        )
+
+    # If we only find one value with this property_, we know this value isn't split
+    # into partial params
+    if len(property_values) == 1:
+        raise TypeError(
+            f"Configuration parameter {property_} for node {node.node_id} does not "
+            "have partials"
+        )
+
+    # If new_value is a dictionary, we need to calculate the full value to send
+    if isinstance(new_value, dict):
+        temp_value = 0
+        # For each property key provided, we bit shift the partial value using the
+        # property_key
+        for property_key, partial_value in new_value.items():
+            value_id = get_value_id(
+                node, CommandClass.CONFIGURATION, property_, property_key=property_key
+            )
+            if value_id not in node.values:
+                raise NotFoundError(
+                    f"Bitmask {property_key} ({hex(property_key)}) not found for "
+                    f"parameter {property_}"
+                )
+            zwave_value = cast(ConfigurationValue, node.values[value_id])
+            partial_value = _validate_and_transform_new_value(
+                zwave_value, partial_value
+            )
+            temp_value += partial_value << partial_param_bit_shift(property_key)
+
+        # To set partial parameters in bulk, we also have to include cached values for
+        # property keys that haven't been specified
+        for property_value in property_values:
+            if property_value.property_key not in new_value:
+                temp_value += cast(
+                    int, property_value.value
+                ) << partial_param_bit_shift(cast(int, property_value.property_key))
+
+        new_value = temp_value
+    else:
+        remaining_value = new_value
+
+        # Break down the bulk value into partial values and validate them against
+        # each partial parameter's metadata
+        for property_key in sorted(
+            [cast(int, value.property_key) for value in property_values], reverse=True
+        ):
+            multiplication_factor = 2 ** partial_param_bit_shift(property_key)
+            partial_value = int(remaining_value / multiplication_factor)
+            remaining_value = remaining_value % multiplication_factor
+            zwave_value = cast(
+                ConfigurationValue,
+                node.values[
```

Comment:
Do we need to guard against a missing value id here?

Reply:
We shouldn't have to because we are looping through property keys that we retrieved from the node, but I just realized there's a better way to do this.